### PR TITLE
[PublishBuildArtifactsV1, DownloadBuildArtifactsV0] Add optional input to store artifacts as tar archives

### PR DESCRIPTION
**Pull request by DaniilShmelev**
**Task name**: PublishBuildArtifactsV1, DownloadBuildArtifactsV0

**Description**: Currently artifacts don't store Unix file permissions. The solution we've decided to implement is to add the option to store artifacts as tar archives, which store permissions.

Since `pathToPublish` is no longer always going to be the path that is going to be uploaded to the server, I've added a new variable `pathToUpload`. `pathToPublish` now has a different purpose:
- `pathToPublish` is a folder or a single file that may be added to a tar archive later
- `pathToUpload` is an actual folder or file that will get uploaded

**Documentation changes required:** Yes

**Added unit tests:** Yes

**Attached related issue:** https://github.com/microsoft/azure-pipelines-tasks/issues/6364

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected - still need more tests. Currently we have these successful pipelines:

❌ [preserve file permissions windows](https://dev.azure.com/v-dshmelev/playground/_build/results?buildId=1019) - fails as expected because `extractTars` input is not allowed in windows
✔ [don't preserve file permissions windows](https://dev.azure.com/v-dshmelev/playground/_build/results?buildId=1020)
✔ [preserve file permissions ubuntu](https://dev.azure.com/v-dshmelev/playground/_build/results?buildId=1021)
✔ [don't preserve file permissions ubuntu](https://dev.azure.com/v-dshmelev/playground/_build/results?buildId=1022)
✔ [preserve file permissions macos](https://dev.azure.com/v-dshmelev/playground/_build/results?buildId=1023)
✔ [don't preserve file permissions macos](https://dev.azure.com/v-dshmelev/playground/_build/results?buildId=1024)
✔ [preserve file permissions self-hosted centos](https://dev.azure.com/v-dshmelev/playground/_build/results?buildId=1013)
✔ [don't preserve file permissions self-hosted centos](https://dev.azure.com/v-dshmelev/playground/_build/results?buildId=1014)

✔ [shows warning if no archives were found](https://dev.azure.com/v-dshmelev/playground/_build/results?buildId=1015)
✔ [download multiple tars](https://dev.azure.com/v-dshmelev/playground/_build/results?buildId=1016)
✔ [download and extract multiple tars ubuntu](https://dev.azure.com/v-dshmelev/playground/_build/results?buildId=1017)
✔ [download and extract multiple tars self-hosted centos](https://dev.azure.com/v-dshmelev/playground/_build/results?buildId=1018)

### CODE REVIEW

File: Tasks/DownloadBuildArtifactsV0/main.ts [hunk 1]
```diff
@@ -332,8 +333,32 @@ async function main(): Promise<void> {
...
+
+                const shouldExtractTars: boolean = tl.getBoolInput('extractTars');
+                if (shouldExtractTars) {
+                    const flatTickets: models.ArtifactDownloadTicket[] = [].concat(...tickets);
+                    const tarArchivesPaths = [] as string[];
+
+                    flatTickets.forEach((ticket: models.ArtifactDownloadTicket) => {
+                        if (
+                            ticket.artifactItem.itemType === models.ItemType.File
+                            && ticket.artifactItem.path.endsWith('.tar')
```
Comment (anatolybolshakov) [inline Tasks/DownloadBuildArtifactsV0/main.ts:347]: Do we need null checks here? Could you please double-check that there properties are always not null?

File: Tasks/DownloadBuildArtifactsV0/main.ts [hunk 2]
```diff
@@ -389,6 +414,36 @@ function configureDownloaderOptions(): engine.ArtifactEngineOptions {
     return downloaderOptions;
 }
 
+function extractTars(downloadPath: string, tarArchivesPaths: string[]): void {
```
Comment (anatolybolshakov) [inline Tasks/DownloadBuildArtifactsV0/main.ts:417]: Nit: could you please add some comments?

File: Tasks/DownloadBuildArtifactsV0/main.ts [hunk 3]
```diff
@@ -389,6 +414,36 @@ function configureDownloaderOptions(): engine.ArtifactEngineOptions {
...
+    const extractedTarsPath: string = path.join(tl.getVariable('Agent.TempDirectory'), 'extracted_tars');
```
Comment (anatolybolshakov) [inline Tasks/DownloadBuildArtifactsV0/main.ts:418]: Please move 'extracted_tars' to constants

File: Tasks/PublishBuildArtifactsV1/task.json [hunk 4]
```diff
@@ -92,6 +92,15 @@
...
             "visibleRule": "ArtifactType = FilePath"
+        },
+        {
+            "name": "StoreAsTar",
+            "type": "boolean",
+            "label": "Tar the artifact before uploading",
+            "defaultValue": "false",
+            "groupName": "advanced",
+            "required": false,
+            "helpMarkDown": "Add all files from the publish path to a tar archive before uploading. This allows you to preserve the UNIX file permissions."
```
Comment (anatolybolshakov) [inline Tasks/PublishBuildArtifactsV1/task.json:103]: Could it make sense to mention similar feature of DownloadBuildArtifacts also here?

File: Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts [hunk 5]
```diff
@@ -43,13 +44,51 @@ let pathToRobocopyPSString = (filePath: string) => {
     return `'${result}'`;
 }
 
+function createTarArchive(filesPath: string, artifactName: string) {
```
Comment (anatolybolshakov) [inline Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts:47]: Nit: would be great to have JS Docs also 😃

File: Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts [hunk 6]
```diff
@@ -71,8 +110,8 @@ async function run() {
             data["containerfolder"] = artifactName;
 
             // add localpath to ##vso command's properties for back compat of old Xplat agent
-            data["localpath"] = pathtoPublish;
-            tl.command("artifact.upload", data, pathtoPublish);
+            data["localpath"] = pathToUpload;
```
Comment (anatolybolshakov) [inline Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts:113]: Please add this update to PR's description

File: Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts [hunk 7]
```diff
@@ -43,13 +44,51 @@ let pathToRobocopyPSString = (filePath: string) => {
...
+    const tarExecResult: tr.IExecSyncResult = tar.execSync();
+
+    if (tarExecResult.error || tarExecResult.code !== 0) {
+        throw new Error(`Couldn't add artifact files to a tar archive: ${tarExecResult.error}`);
+    }
+
+    return outputFilePath;
+}
+
+function getPathToUpload(
```
Comment (anatolybolshakov) [inline Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts:68]: Could it make sense to rename this method? Currently it doesn't only return upload path - for some cases it creates archive, which could be unobvious

File: Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts (hunk 6 above)
Comment (DaniilShmelev) [reply Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts:113]: Updated

File: Tasks/DownloadBuildArtifactsV0/main.ts (hunk 1 above)
Comment (DaniilShmelev) [reply Tasks/DownloadBuildArtifactsV0/main.ts:348]: Yeah, double-checked this: `ArtifactDownloadTicket` type should never have null properties, and the tickets themselves can't be null either

File: Tasks/DownloadBuildArtifactsV0/main.ts (hunk 2 above)
Comment (DaniilShmelev) [reply Tasks/DownloadBuildArtifactsV0/main.ts]: Done

File: Tasks/DownloadBuildArtifactsV0/main.ts (hunk 3 above)
Comment (DaniilShmelev) [reply Tasks/DownloadBuildArtifactsV0/main.ts]: Done

File: Tasks/PublishBuildArtifactsV1/task.json (hunk 4 above)
Comment (DaniilShmelev) [reply Tasks/PublishBuildArtifactsV1/task.json]: Yeah, mentioned matching options in both tasks

File: Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts (hunk 5 above)
Comment (DaniilShmelev) [reply Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts]: Yeah, added them here too

File: Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts (hunk 7 above)
Comment (DaniilShmelev) [reply Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts]: Yeah, renamed it

**Review by anatolybolshakov — APPROVED** [review]
LGTM, thanks!

**Review by EzzhevNikita — APPROVED** [review]

File: Tasks/DownloadBuildArtifactsV0/main.ts [hunk 8]
```diff
@@ -332,8 +334,32 @@ async function main(): Promise<void> {
...
+                    });
+
+                    tl.debug(`Found ${tarArchivesPaths.length} tar archives:\n${tarArchivesPaths.join('\n')}`);
+
+                    if (tarArchivesPaths.length === 0) {
+                        tl.warning(tl.loc('NoTarsFound'));
+                    } else {
+                        extractTars(tarArchivesPaths, downloadPath);
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (DaniilShmelev) [reply Tasks/DownloadBuildArtifactsV0/main.ts]: Thanks, done!

File: Tasks/DownloadBuildArtifactsV0/main.ts [hunk 9]
```diff
@@ -389,6 +415,52 @@ function configureDownloaderOptions(): engine.ArtifactEngineOptions {
...
+function extractTar(tarArchivePath: string, extractedFilesDir: string): void {
+    const tar: tr.ToolRunner = tl.tool(tl.which('tar', true));
+    tl.mkdirP(extractedFilesDir);
+    tar.arg(['xf', tarArchivePath, '--directory', extractedFilesDir]);
+    const tarExecResult: tr.IExecSyncResult = tar.execSync();
+
+    if (tarExecResult.error || tarExecResult.code !== 0) {
+        throw new Error(`Couldn't extract artifact files from a tar archive: ${tarExecResult.error}`);
+    }
+}
```
[the remark this answers is not in the document]
Comment (DaniilShmelev) [reply Tasks/DownloadBuildArtifactsV0/main.ts]: done

**Review by alexander-smolyakov — APPROVED** [review]
LGTM! 

Tested these changes on [ubuntu](https://dev.azure.com/v-alsmo/ArtifactsTest/_build/results?buildId=1110&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=35f53584-79fc-539f-463c-6471d0663a3a) all works as expected.

**Review by alexander-smolyakov — APPROVED** [review]
LGTM, thanks!

Tested these changes on Ubuntu, macOS, and Windows:
- [Ubuntu](https://v-alsmo.visualstudio.com/ArtifactsTest/_build/results?buildId=1126&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9)
- [macOS](https://v-alsmo.visualstudio.com/ArtifactsTest/_build/results?buildId=1127&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=c7b27c9e-27d6-4154-ab7e-52d730239c8f)
- [Windows](https://v-alsmo.visualstudio.com/ArtifactsTest/_build/results?buildId=1123&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=c5799012-d252-50c1-2200-2d285327a1f0) - PublishBuildArtifact
- [Windows](https://v-alsmo.visualstudio.com/ArtifactsTest/_build/results?buildId=1125&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=35bfe49e-764e-5635-9d9f-9d7e536521bf) - DownloadBuildArtifact

**Comment by matskan** [comment]
LGTM 👍

**Pull request merged**